libcore/src/SimulationHelper.cpp

```diff
@@ -0,0 +1,218 @@
+        });
+
+    // pedestrian is in one of the neighboring rooms/subrooms
+    if(currentSubRoom != subroomsConnected.end()) {
+        ped.UpdateRoom(currentSubRoom->second->GetRoomID(), currentSubRoom->second->GetSubRoomID());
+        ped.SetSubRoomUID(currentSubRoom->second->GetUID());
+        return true;
+    } else {
+        ped.UpdateRoom(-1, -1);
+        ped.SetSubRoomUID(-1);
+        return std::nullopt;
+    }
+}
+
+std::vector<Pedestrian *> SimulationHelper::FindPedsReachedFinalGoal(
+    const Building & building,
+    const std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsAtFinalGoal;
+    const auto & goals = building.GetAllGoals();
+
+    std::copy_if(
+        std::begin(peds),
+        std::end(peds),
+        std::inserter(pedsAtFinalGoal, std::end(pedsAtFinalGoal)),
+        [goals](auto const ped) -> bool {
+            return ped->GetFinalDestination() != FINAL_DEST_OUT &&
+                   goals.at(ped->GetFinalDestination())->Contains(ped->GetPos()) &&
+                   goals.at(ped->GetFinalDestination())->GetIsFinalGoal();
+        });
+    return pedsAtFinalGoal;
+}
+
+std::tuple<std::vector<Pedestrian *>, std::vector<Pedestrian *>>
+SimulationHelper::UpdateLocations(const Building & building, const std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsNotRelocated;
+    std::vector<Pedestrian *> pedsChangedRoom;
+
+    // Check for peds, where relocation failed
+    for(auto const ped : peds) {
+        if(auto relocated = UpdateRoom(building, *ped); relocated.has_value()) {
+            if(relocated.value() && ped->ChangedRoom()) {
+                pedsChangedRoom.push_back(ped);
+            }
+        } else {
+            pedsNotRelocated.push_back(ped);
+        }
+    }
+
+    return {pedsChangedRoom, pedsNotRelocated};
+}
+
+std::vector<Pedestrian *> SimulationHelper::FindOutsidePedestrians(
+    const Building & building,
+    std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsOutside;
+    std::vector<Pedestrian *> newPeds;
+
+    std::partition_copy(
+        std::make_move_iterator(std::begin(peds)),
+        std::make_move_iterator(std::end(peds)),
+        std::back_inserter(pedsOutside),
+        std::back_inserter(newPeds),
+        [&building](auto const ped) -> bool {
+            auto transPassed = SimulationHelper::FindPassedDoor(building, *ped);
+            if(!transPassed.has_value()) {
+                return false;
+            }
+
+            //TODO maxDistance should depend on vmax
+            double maxDistance = 0.5;
+            return transPassed.value()->IsExit() && transPassed.value()->IsOpen() &&
+                   transPassed.value()->DistTo(ped->GetPos()) < maxDistance;
+        });
+    peds = std::move(newPeds);
+    return pedsOutside;
+}
+
+void SimulationHelper::UpdateFlowAtDoors(
+    Building & building,
+    const std::vector<Pedestrian *> & peds)
```

Comment:
Maybe this should be called `peds_changed_room` to be clear that this function is called for some and not for all pedestrians...

Reply:
Changed the name